program-runtime/src/sysvar_cache.rs

```diff
@@ -312,3 +349,39 @@ pub mod get_sysvar_with_account_check {
         invoke_context.get_sysvar_cache().get_last_restart_slot()
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use {super::*, test_case::test_case};
+
+    // sysvar cache provides the full account data of a sysvar
+    // the setters MUST NOT be changed to serialize an object representation
+    // it is required that the syscall be able to access the full buffer as it exists onchain
+    #[test_case(Clock::default(); "clock")]
+    #[test_case(EpochSchedule::default(); "epoch_schedule")]
+    #[test_case(EpochRewards::default(); "epoch_rewards")]
+    #[test_case(Rent::default(); "rent")]
+    #[test_case(SlotHashes::default(); "slot_hashes")]
+    #[test_case(StakeHistory::default(); "stake_history")]
+    #[test_case(LastRestartSlot::default(); "last_restart_slot")]
+    fn test_sysvar_cache_preserves_bytes<T: Sysvar>(_: T) {
+        let id = T::id();
+        let size = T::size_of();
+        let in_buf = vec![0; size];
+        let mut out_buf = vec![0xff; size];
+
+        let mut sysvar_cache = SysvarCache::default();
+        sysvar_cache.fill_missing_entries(|pubkey, callback| {
+            if *pubkey == id {
+                callback(&in_buf)
+            }
+        });
+        let sysvar_cache = sysvar_cache;
+
+        sysvar_cache
+            .read_sysvar_into(&id, size, 0, &mut out_buf)
+            .unwrap();
+
+        assert_eq!(out_buf, in_buf);
+    }
```

Comment:
I might be missing something here, but since the docs say this test is supposed to assert the sysvar cache is keeping all of the bytes from the account data, should it be able to check & handle a case where, say, the account data is larger than `T::size_of()`?

Reply:
thats an interesting point i hadnt considered. i guess we can also cover that case by making `size` bigger than it needs to be. the specific case i was trying to cover was just that if you roundtrip a `StakeHistory` or `SlotHashes` that has less than `MAX_ENTRIES` through bincode it truncates the `Vec`

Reply:
incidentally i realize this means we do in fact want to use vectors rather than arrays in the cache (remember discussing this in the past) because we have no access to the actual size of the onchain data in code

Reply:
Yeah, I was going to comment on the vectors vs. arrays in this PR as well, but you're right I think it may have to be vectors for that reason.

> i guess we can also cover that case by making size bigger than it needs to be.

Yeah I think we definitely should have this in the tests somewhere, since FD was pretty big on this point in the discussions.